sklearn/metrics/_dist_metrics.pyx.tp

```diff
@@ -770,11 +838,15 @@ cdef class DistanceMetric{{name_suffix}}:
         Y_is_sparse = issparse(Y)
 
         if not X_is_sparse and not Y_is_sparse:
+            X = np.asarray(X, order='C')
+            Y = np.asarray(Y, order='C')
             return self._pairwise_dense_dense(X, Y)
         if X_is_sparse and Y_is_sparse:
             return self._pairwise_sparse_sparse(X, Y)
         if X_is_sparse and not Y_is_sparse:
+            Y = np.asarray(Y, order='C')
```

Comment:
Nit: If we want to be consistent with `_pairwise_dense_dense`, then this casting should be in `_pairwise_sparse_dense`. We end up with the following in `_pairwise_sparse_dense`:

```python
def _pairwise_sparse_dense(X, Y): 
    ... 
    const {{INPUT_DTYPE_t}}[:, ::1] Y_data =  np.asarray(Y, dtype={{INPUT_DTYPE}}, order="C")
    ...

    for i1 in range(n_X):
        for i2 in range(n_Y):
            x2_data = &Y_data[0, 0] + i2 * n_features
```

The same idea for `_pairwise_dense_sparse`. This is not blocking since `pairwise` is only used for testing.